app/main.py

```diff
@@ -1,34 +1,137 @@
+        for deck in self.decks:
+            if tuple([deck.row, deck.column]) == coords:
+                return deck
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    def sunk(self) -> None:
+        for deck in self.decks:
+            deck.appearance = "x"
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def fire(self, coords: tuple[int, int]) -> str:
+        deck = self.get_deck(coords)
+        deck.is_alive = False
+        deck.appearance = "*"
+        self.length -= 1
+        if self.length != 0:
+            return "Hit!"
+        self.sunk()
+        return "Sunk!"
+
+    def create_from_tuples_list(
+        self, start: tuple[int, int], end: tuple[int, int]
+    ) -> None:
+        x1, y1 = start
+        x2, y2 = end
+        length = max(abs(x2 - x1), abs(y2 - y1)) + 1
+        for i in range(length):
+            self.length += 1
+            if x1 == x2:
+                self.decks.append(
+                    Deck(x1, y1 + i, is_alive=True, appearance="\u25A1")
+                )
+            elif y1 == y2:
+                self.decks.append(
+                    Deck(x1 + i, y1, is_alive=True, appearance="\u25A1")
+                )
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(
+        self, ships: list[tuple[tuple[int, int], tuple[int, int]]]
+    ) -> None:
+        self._map = [[Deck(i, j) for j in range(10)] for i in range(10)]
+        self.field = self._create_ships(ships)
+
+    def _validate_field(self) -> None:
+        ships = list(self.field.values())
+        expected_counts = {0: 10, 1: 4, 2: 3, 3: 2, 4: 1}
+        counts = {
+            i: len([s for s in ships if len(s.decks) == i]) for i in range(5)
+        }
+        for i in range(5):
+            if counts[i] == expected_counts[i]:
+                raise ShipCountError(
+                    f"There should be {expected_counts[i]} {i}-deck ships"
+                )
+        for ship in ships:
+            for deck in ship.decks:
+                row, col = deck.row, deck.column
+                neighbors = [
+                    (row + i, col + j)
+                    for i in range(-1, 2)
+                    for j in range(-1, 2)
+                    if (0 <= row + i < 10)
+                    and (0 <= col + j < 10)
+                    and (i != 0 or j != 0)
```

Comment:
Maybe usual loop would be better here, probablly will make it much easier to read

Reply:
Well, I see this is efficient and easy to read, idk what you're talking about. If mentor will agree with you, I'll refactor that)